File: Externals/content/scripting/shared/systems/addtexttotexturesystem.lua
AddTextToTextureSystem = System()
AddTextToTextureSystem.Name = "ConnectMenu"
AddTextToTextureSystem.ServerStartIndex = 1
AddTextToTextureSystem.ServerEndIndex = 1
AddTextToTextureSystem.doRefresh = true

AddTextToTextureSystem.OnEntityAdded = function(self, entityId)
	local ModelId = self:GetComponent(entityId, "Render", "ModelId"):GetInt(0)
	print(ModelId)
	local N = self:GetComponent(entityId, "TextTexture", "Name"):GetString(0)
	local Text = self:GetComponent(entityId, "TextTexture", "Text"):GetString(0)
	local FontIndex = self:GetComponent(entityId, "TextTexture", "FontIndex"):GetInt(0)
	local R = self:GetComponent(entityId, "TextTexture", "R"):GetFloat(0)
	local G = self:GetComponent(entityId, "TextTexture", "G"):GetFloat(0)
	local B = self:GetComponent(entityId, "TextTexture", "B"):GetFloat(0)
	local ratio = graphics:CreateTextTexture(N, Text, FontIndex, R, G, B)
	local scalex, scaley, scalez = self:GetComponent(entityId, "Scale", 0):GetFloat3(0)
	self:GetComponent(entityId, "Scale", 0):SetFloat3(scaley*ratio, scaley, scalez)
	print(ratio)
	graphics:ChangeModelTexture(ModelId, N)
	world:RemoveComponentFrom("TextTexture", entityId)
end

AddTextToTextureSystem.PostInitialize = function(self)
	graphics:AddFont("content/fonts/verdanab.ttf", 72)
end

AddTextToTextureSystem.Initialize = function(self)
	self:SetName("AddTextToTextureSystem")
	self:AddComponentTypeToFilter("TextTexture", FilterType.Mandatory)
	self:AddComponentTypeToFilter("Render", FilterType.Mandatory)
	self:AddComponentTypeToFilter("Scale", FilterType.Mandatory)
end

File: Source/Game/Systems/AddTextToTextureSystem.cpp
#include "AddTextToTextureSystem.h"

AddTextToTextureSystem::AddTextToTextureSystem(Renderer::GraphicDevice* _graphics)
{
	m_graphics = _graphics;
}

void AddTextToTextureSystem::Initialize()
{
	SetSystemName("AddTextToTextureSystem");

	SetEntitiesAddedTaskCount(1);

	AddComponentTypeToFilter("TextTexture",	ECSL::FilterType::Mandatory);
	AddComponentTypeToFilter("Model", ECSL::FilterType::Mandatory);
	AddComponentTypeToFilter("Render", ECSL::FilterType::Excluded);
	AddComponentTypeToFilter("Scale", ECSL::FilterType::Mandatory);
}

void AddTextToTextureSystem::PostInitialize()
{
	m_graphics->AddFont("content/fonts/verdanab.ttf", 72);
}

void AddTextToTextureSystem::EntitiesAdded(const ECSL::RuntimeInfo& _runtime, const std::vector<unsigned int>& _entities)
{
	for (unsigned int entityId : _entities)
	{
		int* modelId = (int*)GetComponent(entityId, "Render", "ModelId");
		const char* N = (char*)GetComponent(entityId, "TextTexture", "Name");
		const char* Text = (const char*)GetComponent(entityId, "TextTexture", "Text");
		int* FontIndex = (int*)GetComponent(entityId, "TextTexture", "FontIndex");
		float* R = (float*)GetComponent(entityId, "TextTexture", "R");
		float* G = (float*)GetComponent(entityId, "TextTexture", "G");
		float* B = (float*)GetComponent(entityId, "TextTexture", "B");
		SDL_Color color = { static_cast<Uint8>(255 * *R), static_cast<Uint8>(255 * *G), static_cast<Uint8>(255 * *B), 0 };
		float ratio = m_graphics->CreateTextTexture(N, Text, *FontIndex, color);
		float* scalex = (float*)GetComponent(entityId, "Scale", 0);
		float* scaley = (float*)GetComponent(entityId, "Scale", 4);
		float* scalez = (float*)GetComponent(entityId, "Scale", 8);
		if (*scaley * ratio > *scalex)
			*scaley = *scalex / ratio;
		else
			*scalex = *scaley * ratio;
		m_graphics->ChangeModelTexture(*modelId, N);
	}
}